### PR TITLE
LPD-31929 - Address ERC gaps in Headless APIs for categoryDisplayPages

### PR DESCRIPTION
**Pull request by matyaswollner**
https://liferay.atlassian.net/browse/LPD-31929

### CODE REVIEW

**Comment by matyaswollner** [comment]
ci:test:sf

**Comment by smottal** [comment]
ci:test:relevant

**Comment by matyaswollner** [comment]
ci:test:sf

**Comment by matyaswollner** [comment]
ci:test:relevant

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java [hunk 1]
```diff
@@ -168,11 +183,26 @@ public CategoryDisplayPage postChannelIdCategoryDisplayPage(
...
 			_commerceChannelLocalService.getCommerceChannel(id);
 
+		long categoryId = GetterUtil.getLong(
+			categoryDisplayPage.getCategoryId());
+
+		if (categoryId == 0) {
+			AssetCategory assetCategory =
+				_assetCategoryService.fetchCategoryByExternalReferenceCode(
+					categoryDisplayPage.getCategoryExternalReferenceCode(),
+					commerceChannel.getSiteGroupId());
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java:193]: here it should be the company group id

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java [hunk 2]
```diff
@@ -131,12 +133,25 @@ public CategoryDisplayPage patchCategoryDisplayPage(
...
 		}
 
+		long categoryId = GetterUtil.getLong(
+			categoryDisplayPage.getCategoryId());
+
+		if (categoryId == 0) {
+			AssetCategory assetCategory =
+				_assetCategoryService.fetchCategoryByExternalReferenceCode(
+					categoryDisplayPage.getCategoryExternalReferenceCode(),
+					cpDisplayLayout.getGroupId());
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java:143]: here should be company gruop id

**Review by andrea-ale-sbarra — CHANGES_REQUESTED** [review]
I think you should use the company group id because the category we can use are from the Global scope only

**Comment by matyaswollner** [comment]
ci:test:sf

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/dto/v1_0/converter/CategoryDisplayPageDTOConverter.java [hunk 3]
```diff
@@ -40,13 +42,28 @@ public CategoryDisplayPage toDTO(DTOConverterContext dtoConverterContext)
...
+						AssetCategory assetCategory =
+							_assetCategoryLocalService.fetchAssetCategory(
+								cpDisplayLayout.getClassPK());
+
+						if (assetCategory == null) {
+							return null;
+						}
+
+						return assetCategory.getExternalReferenceCode();
+					});
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/dto/v1_0/converter/CategoryDisplayPageDTOConverter.java:56]: please set also the GroupERC of the category

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java [hunk 4]
```diff
@@ -131,12 +136,33 @@ public CategoryDisplayPage patchCategoryDisplayPage(
...
+		long categoryId = GetterUtil.getLong(
+			categoryDisplayPage.getCategoryId());
+
+		if (categoryId == 0) {
+			Group group = _groupService.fetchGroupByExternalReferenceCode(
+				categoryDisplayPage.getGroupExternalReferenceCode(),
+				contextCompany.getCompanyId());
+
+			if (group == null) {
+				throw new NoSuchGroupException();
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java:148]: this doesn't look right

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java [hunk 5]
```diff
@@ -165,14 +191,37 @@ public CategoryDisplayPage postChannelIdCategoryDisplayPage(
...
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java:203]: I'm not sure we should be throwing exception in cases like this. can you please check with other APIs?

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java [hunk 6]
```diff
@@ -165,14 +191,37 @@ public CategoryDisplayPage postChannelIdCategoryDisplayPage(
 			Long id, CategoryDisplayPage categoryDisplayPage)
 		throws Exception {
 
+		long categoryId = GetterUtil.getLong(
+			categoryDisplayPage.getCategoryId());
+
+		if (categoryId == 0) {
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-channel-impl/src/main/java/com/liferay/headless/commerce/admin/channel/internal/resource/v1_0/CategoryDisplayPageResourceImpl.java:197]: if it is not 0 did you make sure this is an actual category id?

**Review by andrea-ale-sbarra — CHANGES_REQUESTED** [review]

**Comment by smottal** [comment]
Resent here: https://github.com/liferay-commerce/liferay-portal/pull/5339

@matyaswollner please have a look to my changes. Thx

**Pull request closed without merging**